### DISCUSSION TITLE
strimzi-cluster-operator pod remains in CrashLoopBackOff and keeps restarting

### DISCUSSION THREAD

**Discussion by natu92** [strimzi/strimzi-kafka-operator#6229] [Q&A]
**.yml used:** strimzi-cluster-operator-0.27.0.yaml

**kubectl get pods -n kafka:**
NAME                                          READY   STATUS             RESTARTS         AGE
my-cluster-entity-operator-868866f44b-b87q2   3/3     Running            0                30h
my-cluster-kafka-0                            1/1     Running            0                32h
my-cluster-zookeeper-0                        1/1     Running            0                32h
my-cluster-zookeeper-1                        1/1     Running            0                32h
my-cluster-zookeeper-2                        1/1     Running            0                32h
strimzi-cluster-operator-7d96cbff57-s47ns     0/1     CrashLoopBackOff   361 (4m8s ago)   32h

**On describing the pod:**
$ kubectl describe pod strimzi-cluster-operator-7d96cbff57-s47ns -n kafka
Name:         strimzi-cluster-operator-7d96cbff57-s47ns
Namespace:    kafka
Priority:     0
Node:         docker-desktop/192.168.65.4
Start Time:   Mon, 17 Jan 2022 06:24:11 +0000
Labels:       name=strimzi-cluster-operator
              pod-template-hash=7d96cbff57
              strimzi.io/kind=cluster-operator
Annotations:  <none>
Status:       Running
IP:           10.1.0.6
IPs:
  IP:           10.1.0.6
Controlled By:  ReplicaSet/strimzi-cluster-operator-7d96cbff57
Containers:
  strimzi-cluster-operator:
    Container ID:  docker://717b2da4e1ea776becac09d3af45da011ebd25c49702508ea53d558cc102a827
    Image:         quay.io/strimzi/operator:0.27.0
    Image ID:      docker-pullable://quay.io/strimzi/operator@sha256:b8ec0cf995659c417e0b849991f66adc31cf0f2484306fe9d07f382074f4fb7b
    Port:          8080/TCP
    Host Port:     0/TCP
    Args:
      /opt/strimzi/bin/cluster_operator_run.sh
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Tue, 18 Jan 2022 14:32:04 +0000
      Finished:     Tue, 18 Jan 2022 14:32:33 +0000
    Ready:          False
    Restart Count:  361
    Limits:
      cpu:     1
      memory:  384Mi
    Requests:
      cpu:      200m
      memory:   384Mi
    Liveness:   http-get http://:http/healthy delay=10s timeout=1s period=30s #success=1 #failure=3
    Readiness:  http-get http://:http/ready delay=10s timeout=1s period=30s #success=1 #failure=3
    Environment:
      STRIMZI_NAMESPACE:                                  kafka (v1:metadata.namespace)
      STRIMZI_FULL_RECONCILIATION_INTERVAL_MS:            120000
      STRIMZI_OPERATION_TIMEOUT_MS:                       300000
      STRIMZI_DEFAULT_TLS_SIDECAR_ENTITY_OPERATOR_IMAGE:  quay.io/strimzi/kafka:0.27.0-kafka-3.0.0
      STRIMZI_DEFAULT_KAFKA_EXPORTER_IMAGE:               quay.io/strimzi/kafka:0.27.0-kafka-3.0.0
      STRIMZI_DEFAULT_CRUISE_CONTROL_IMAGE:               quay.io/strimzi/kafka:0.27.0-kafka-3.0.0
      STRIMZI_DEFAULT_TLS_SIDECAR_CRUISE_CONTROL_IMAGE:   quay.io/strimzi/kafka:0.27.0-kafka-3.0.0
      STRIMZI_KAFKA_IMAGES:                               2.8.0=quay.io/strimzi/kafka:0.27.0-kafka-2.8.0
                                                          2.8.1=quay.io/strimzi/kafka:0.27.0-kafka-2.8.1
                                                          3.0.0=quay.io/strimzi/kafka:0.27.0-kafka-3.0.0

      STRIMZI_KAFKA_CONNECT_IMAGES:                       2.8.0=quay.io/strimzi/kafka:0.27.0-kafka-2.8.0
                                                          2.8.1=quay.io/strimzi/kafka:0.27.0-kafka-2.8.1
                                                          3.0.0=quay.io/strimzi/kafka:0.27.0-kafka-3.0.0

      STRIMZI_KAFKA_MIRROR_MAKER_IMAGES:                  2.8.0=quay.io/strimzi/kafka:0.27.0-kafka-2.8.0
                                                          2.8.1=quay.io/strimzi/kafka:0.27.0-kafka-2.8.1
                                                          3.0.0=quay.io/strimzi/kafka:0.27.0-kafka-3.0.0

      STRIMZI_KAFKA_MIRROR_MAKER_2_IMAGES:                2.8.0=quay.io/strimzi/kafka:0.27.0-kafka-2.8.0
                                                          2.8.1=quay.io/strimzi/kafka:0.27.0-kafka-2.8.1
                                                          3.0.0=quay.io/strimzi/kafka:0.27.0-kafka-3.0.0

      STRIMZI_DEFAULT_TOPIC_OPERATOR_IMAGE:               quay.io/strimzi/operator:0.27.0
      STRIMZI_DEFAULT_USER_OPERATOR_IMAGE:                quay.io/strimzi/operator:0.27.0
      STRIMZI_DEFAULT_KAFKA_INIT_IMAGE:                   quay.io/strimzi/operator:0.27.0
      STRIMZI_DEFAULT_KAFKA_BRIDGE_IMAGE:                 quay.io/strimzi/kafka-bridge:0.21.2
      STRIMZI_DEFAULT_JMXTRANS_IMAGE:                     quay.io/strimzi/jmxtrans:0.27.0
      STRIMZI_DEFAULT_KANIKO_EXECUTOR_IMAGE:              quay.io/strimzi/kaniko-executor:0.27.0
      STRIMZI_DEFAULT_MAVEN_BUILDER:                      quay.io/strimzi/maven-builder:0.27.0
      STRIMZI_OPERATOR_NAMESPACE:                         kafka (v1:metadata.namespace)
      STRIMZI_FEATURE_GATES:
    Mounts:
      /opt/strimzi/custom-config/ from co-config-volume (rw)
      /tmp from strimzi-tmp (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-xl494 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  strimzi-tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  1Mi
  co-config-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      strimzi-cluster-operator
    Optional:  false
  kube-api-access-xl494:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason   Age                    From     Message
  ----     ------   ----                   ----     -------
  Normal   Pulled   17m (x360 over 32h)    kubelet  Container image "quay.io/strimzi/operator:0.27.0" already present on machine
  Warning  BackOff  2m1s (x8293 over 32h)  kubelet  Back-off restarting failed container

Strangely, this issue appears to happen with latest docker version (20.10.11) with Hyper-V enabled. However, it doesn't occur when WSL2 is enabled. Please share pointers, if any.

**DOCKER INFO:**
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.7.1)
  compose: Docker Compose (Docker Inc., v2.2.1)
  scan: Docker Scan (Docker Inc., v0.14.0)

Server:
 Containers: 110
  Running: 45
  Paused: 0
  Stopped: 65
 Images: 21
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 9
 Total Memory: 93.55GiB
 Name: docker-desktop
 ID: J5L2:CTXP:TTP7:X7GK:FL4V:4PUW:27DG:BDEQ:RDZO:CEU6:5S4W:GEZ2
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

**Comment by scholzj** [comment]
The output you shared is quite hard to read because it is not properly formatted. You should check the Kubernetes events and logs to see why was it restarted. One thing to check would be also if it has enough resources available.

**Reply by natu92** [reply]
The restarts happened with exit code 137 which I found to be OOM. Also, I have allocated 95 GB RAM for this setup. 

This issue appear to happen only with latest docker versions (with hyper v) . Same .yml works fine with wsl2 with no change in resource limit.

**Reply by natu92** [reply]
I found a disimiliarity between hyper V and wsl2 setup using docker info command. But, not sure if it has anything to do with this issue. 

In hyper V:
Cgroup Driver: cgroupfs
Cgroup Version: 2

In wsl2: 
Cgroup Driver: cgroupfs
Cgroup Version: 1

**Reply by scholzj** [reply]
Well, I guess 95GB of RAM is the total machine setup. Do you have memory request for the operator it self? Maybe the hypervisors behave differently in what they actually give to the operator.
